# Ensemble Learning

## Initial Imports

In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter
# Initial imports
from sklearn import tree
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
from sklearn.metrics import balanced_accuracy_score


# Needed for decision tree visualization
import pydotplus
from IPython.display import Image

In [3]:
from sklearn.metrics import balanced_accuracy_score

from imblearn.metrics import classification_report_imbalanced

## Read the CSV and Perform Basic Data Cleaning

In [4]:
# Load the data
file_path = Path('Resources/LoanStats_2019Q1.csv')
df = pd.read_csv(file_path)

# Preview the data
df.head()

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0    10500.0    0.1719       375.35           RENT     66000.0   
1    25000.0    0.2000       929.09       MORTGAGE    105000.0   
2    20000.0    0.2000       529.88       MORTGAGE     56000.0   
3    10000.0    0.1640       353.55           RENT     92000.0   
4    22000.0    0.1474       520.39       MORTGAGE     52000.0   

  verification_status   issue_d loan_status pymnt_plan    dti  ...  \
0     Source Verified  Mar-2019    low_risk          n  27.24  ...   
1            Verified  Mar-2019    low_risk          n  20.23  ...   
2            Verified  Mar-2019    low_risk          n  24.26  ...   
3            Verified  Mar-2019    low_risk          n  31.44  ...   
4        Not Verified  Mar-2019    low_risk          n  18.76  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            85.7             100.0                   0.0        0.0   
1            91.2              50.0                   1.0        0.0   
2            66.7              50.0                   0.0        0.0   
3           100.0              50.0                   1.0        0.0   
4           100.0               0.0                   0.0        0.0   

   tot_hi_cred_lim  total_bal_ex_mort total_bc_limit  \
0          65687.0            38199.0         2000.0   
1         271427.0            60641.0        41200.0   
2          60644.0            45684.0         7500.0   
3          99506.0            68784.0        19700.0   
4         219750.0            25919.0        27600.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     61987.0              N                     N  
1                     49197.0              N                     N  
2                     43144.0              N                     N  
3                     76506.0              N                     N  
4                     20000.0              N                     N  

[5 rows x 86 columns]

In [5]:
df.describe()

loan_amnt      int_rate   installment    annual_inc           dti  \
count  68817.000000  68817.000000  68817.000000  6.881700e+04  68817.000000   
mean   16677.594562      0.127718    480.652863  8.821371e+04     21.778153   
std    10277.348590      0.048130    288.062432  1.155800e+05     20.199244   
min     1000.000000      0.060000     30.890000  4.000000e+01      0.000000   
25%     9000.000000      0.088100    265.730000  5.000000e+04     13.890000   
50%    15000.000000      0.118000    404.560000  7.300000e+04     19.760000   
75%    24000.000000      0.155700    648.100000  1.040000e+05     26.660000   
max    40000.000000      0.308400   1676.230000  8.797500e+06    999.000000   

        delinq_2yrs  inq_last_6mths      open_acc       pub_rec  \
count  68817.000000    68817.000000  68817.000000  68817.000000   
mean       0.217766        0.497697     12.587340      0.126030   
std        0.718367        0.758122      6.022869      0.336797   
min        0.000000        0.000000      2.000000      0.000000   
25%        0.000000        0.000000      8.000000      0.000000   
50%        0.000000        0.000000     11.000000      0.000000   
75%        0.000000        1.000000     16.000000      0.000000   
max       18.000000        5.000000     72.000000      4.000000   

           revol_bal  ...  num_tl_90g_dpd_24m  num_tl_op_past_12m  \
count   68817.000000  ...        68817.000000        68817.000000   
mean    17604.142828  ...            0.052138            2.219423   
std     21835.880400  ...            0.390633            1.897432   
min         0.000000  ...            0.000000            0.000000   
25%      6293.000000  ...            0.000000            1.000000   
50%     12068.000000  ...            0.000000            2.000000   
75%     21735.000000  ...            0.000000            3.000000   
max    587191.000000  ...           18.000000           19.000000   

       pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
count    68817.000000      68817.000000          68817.000000    68817.0   
mean        95.057627         30.626217              0.125972        0.0   
std          8.326426         33.631463              0.336732        0.0   
min         20.000000          0.000000              0.000000        0.0   
25%         93.000000          0.000000              0.000000        0.0   
50%        100.000000         20.000000              0.000000        0.0   
75%        100.000000         50.000000              0.000000        0.0   
max        100.000000        100.000000              4.000000        0.0   

       tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
count     6.881700e+04       6.881700e+04    68817.000000   
mean      2.100332e+05       6.133843e+04    29734.128558   
std       1.928088e+05       5.738798e+04    26795.394232   
min       3.600000e+03       2.350000e+02      100.000000   
25%       6.697700e+04       2.650300e+04    11600.000000   
50%       1.467100e+05       4.535700e+04    22100.000000   
75%       3.036400e+05       7.657000e+04    39300.000000   
max       3.292782e+06       1.295455e+06   509400.000000   

       total_il_high_credit_limit  
count                6.881700e+04  
mean                 5.572240e+04  
std                  5.095845e+04  
min                  1.270000e+02  
25%                  2.288000e+04  
50%                  4.200000e+04  
75%                  7.249900e+04  
max                  1.426964e+06  

[8 rows x 76 columns]

In [6]:
df.columns


Index(['loan_amnt', 'int_rate', 'installment', 'home_ownership', 'annual_inc',
       'verification_status', 'issue_d', 'loan_status', 'pymnt_plan', 'dti',
       'delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal',
       'total_acc', 'initial_list_status', 'out_prncp', 'out_prncp_inv',
       'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'next_pymnt_d', 'collections_12_mths_ex_med',
       'policy_code', 'application_type', 'acc_now_delinq', 'tot_coll_amt',
       'tot_cur_bal', 'open_acc_6m', 'open_act_il', 'open_il_12m',
       'open_il_24m', 'mths_since_rcnt_il', 'total_bal_il', 'il_util',
       'open_rv_12m', 'open_rv_24m', 'max_bal_bc', 'all_util',
       'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_si

## Split the Data into Training and Testing

In [7]:
# Create our features
X = df.copy()
X.drop("loan_status", axis=1, inplace=True)
X.head()

# Create our target
y = df["loan_status"]
y.head()

0    low_risk
1    low_risk
2    low_risk
3    low_risk
4    low_risk
Name: loan_status, dtype: object

In [8]:
X.describe()

loan_amnt      int_rate   installment    annual_inc           dti  \
count  68817.000000  68817.000000  68817.000000  6.881700e+04  68817.000000   
mean   16677.594562      0.127718    480.652863  8.821371e+04     21.778153   
std    10277.348590      0.048130    288.062432  1.155800e+05     20.199244   
min     1000.000000      0.060000     30.890000  4.000000e+01      0.000000   
25%     9000.000000      0.088100    265.730000  5.000000e+04     13.890000   
50%    15000.000000      0.118000    404.560000  7.300000e+04     19.760000   
75%    24000.000000      0.155700    648.100000  1.040000e+05     26.660000   
max    40000.000000      0.308400   1676.230000  8.797500e+06    999.000000   

        delinq_2yrs  inq_last_6mths      open_acc       pub_rec  \
count  68817.000000    68817.000000  68817.000000  68817.000000   
mean       0.217766        0.497697     12.587340      0.126030   
std        0.718367        0.758122      6.022869      0.336797   
min        0.000000        0.000000      2.000000      0.000000   
25%        0.000000        0.000000      8.000000      0.000000   
50%        0.000000        0.000000     11.000000      0.000000   
75%        0.000000        1.000000     16.000000      0.000000   
max       18.000000        5.000000     72.000000      4.000000   

           revol_bal  ...  num_tl_90g_dpd_24m  num_tl_op_past_12m  \
count   68817.000000  ...        68817.000000        68817.000000   
mean    17604.142828  ...            0.052138            2.219423   
std     21835.880400  ...            0.390633            1.897432   
min         0.000000  ...            0.000000            0.000000   
25%      6293.000000  ...            0.000000            1.000000   
50%     12068.000000  ...            0.000000            2.000000   
75%     21735.000000  ...            0.000000            3.000000   
max    587191.000000  ...           18.000000           19.000000   

       pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
count    68817.000000      68817.000000          68817.000000    68817.0   
mean        95.057627         30.626217              0.125972        0.0   
std          8.326426         33.631463              0.336732        0.0   
min         20.000000          0.000000              0.000000        0.0   
25%         93.000000          0.000000              0.000000        0.0   
50%        100.000000         20.000000              0.000000        0.0   
75%        100.000000         50.000000              0.000000        0.0   
max        100.000000        100.000000              4.000000        0.0   

       tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
count     6.881700e+04       6.881700e+04    68817.000000   
mean      2.100332e+05       6.133843e+04    29734.128558   
std       1.928088e+05       5.738798e+04    26795.394232   
min       3.600000e+03       2.350000e+02      100.000000   
25%       6.697700e+04       2.650300e+04    11600.000000   
50%       1.467100e+05       4.535700e+04    22100.000000   
75%       3.036400e+05       7.657000e+04    39300.000000   
max       3.292782e+06       1.295455e+06   509400.000000   

       total_il_high_credit_limit  
count                6.881700e+04  
mean                 5.572240e+04  
std                  5.095845e+04  
min                  1.270000e+02  
25%                  2.288000e+04  
50%                  4.200000e+04  
75%                  7.249900e+04  
max                  1.426964e+06  

[8 rows x 76 columns]

In [9]:
X = pd.get_dummies(X)
X

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0        10500.0    0.1719       375.35     66000.0  27.24          0.0   
1        25000.0    0.2000       929.09    105000.0  20.23          0.0   
2        20000.0    0.2000       529.88     56000.0  24.26          0.0   
3        10000.0    0.1640       353.55     92000.0  31.44          0.0   
4        22000.0    0.1474       520.39     52000.0  18.76          0.0   
...          ...       ...          ...         ...    ...          ...   
68812    10000.0    0.1502       346.76     26000.0   9.60          0.0   
68813    12000.0    0.2727       368.37     63000.0  29.07          0.0   
68814     5000.0    0.1992       185.62     52000.0  14.86          0.0   
68815    40000.0    0.0646      1225.24    520000.0   9.96          0.0   
68816    16000.0    0.1131       350.36     72000.0   7.02          2.0   

       inq_last_6mths  open_acc  pub_rec  revol_bal  ...  issue_d_Mar-2019  \
0                 0.0       8.0      0.0     1609.0  ...                 1   
1                 0.0      17.0      1.0    18368.0  ...                 1   
2                 0.0       8.0      0.0    13247.0  ...                 1   
3                 1.0      10.0      1.0    17996.0  ...                 1   
4                 1.0      14.0      0.0     9091.0  ...                 1   
...               ...       ...      ...        ...  ...               ...   
68812             0.0       9.0      0.0     2684.0  ...                 0   
68813             0.0       8.0      0.0    13314.0  ...                 0   
68814             0.0       5.0      1.0     3715.0  ...                 0   
68815             1.0      21.0      0.0    59529.0  ...                 0   
68816             0.0      12.0      1.0    11882.0  ...                 0   

       pymnt_plan_n  initial_list_status_f  initial_list_status_w  \
0                 1                      0                      1   
1                 1                      0                      1   
2                 1                      0                      1   
3                 1                      0                      1   
4                 1                      0                      1   
...             ...                    ...                    ...   
68812             1                      0                      1   
68813             1                      0                      1   
68814             1                      0                      1   
68815             1                      1                      0   
68816             1                      0                      1   

       next_pymnt_d_Apr-2019  next_pymnt_d_May-2019  \
0                          0                      1   
1                          0                      1   
2                          0                      1   
3                          0                      1   
4                          0                      1   
...                      ...                    ...   
68812                      0                      1   
68813                      0                      1   
68814                      0                      1   
68815                      0                      1   
68816                      0                      1   

       application_type_Individual  application_type_Joint App  \
0                                1                           0   
1                                1                           0   
2                                1                           0   
3                                1                           0   
4                                1                           0   
...                            ...                         ...   
68812                            1                           0   
68813                            1                           0   
68814                            1                           0   
68815              

In [10]:
# Check the balance of our target values
y.value_counts()

low_risk     68470
high_risk      347
Name: loan_status, dtype: int64

In [11]:
from sklearn.model_selection import train_test_split

# Split the X and y into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 78) 

## Data Pre-Processing

Scale the training and testing data using the `StandardScaler` from `sklearn`. Remember that when scaling the data, you only scale the features data (`X_train` and `X_testing`).

In [12]:
# Create the StandardScaler instance
scaler = StandardScaler()


In [13]:
# Fit the Standard Scaler with the training data
# When fitting scaling functions, only train on the training dataset
X_scaler =scaler.fit(X_train)

In [14]:
# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Ensemble Learners

In this section, you will compare two ensemble algorithms to determine which algorithm results in the best performance. You will train a Balanced Random Forest Classifier and an Easy Ensemble classifier . For each algorithm, be sure to complete the folliowing steps:

1. Train the model using the training data. 
2. Calculate the balanced accuracy score from sklearn.metrics.
3. Display the confusion matrix from sklearn.metrics.
4. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.
5. For the Balanced Random Forest Classifier only, print the feature importance sorted in descending order (most important feature to least important) along with the feature score

Note: Use a random state of 1 for each algorithm to ensure consistency between tests

### Balanced Random Forest Classifier

In [15]:
from imblearn.ensemble import BalancedRandomForestClassifier


In [16]:
# Resample the training data with the BalancedRandomForestClassifier
brf = BalancedRandomForestClassifier(100, random_state = 1)
brf.fit(X_train_scaled,y_train)

BalancedRandomForestClassifier(random_state=1)

In [17]:
y_predicted = brf.predict(X_test_scaled)

In [18]:
# Calculated the balanced accuracy score
balanced_accuracy_score(y_test, y_predicted)*100

77.90765474505778

In [19]:
# Display the confusion matrix
confusion_matrix(y_test, y_predicted)

array([[   62,    30],
       [ 1981, 15132]])

In [20]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_predicted))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.03      0.67      0.88      0.06      0.77      0.58        92
   low_risk       1.00      0.88      0.67      0.94      0.77      0.61     17113

avg / total       0.99      0.88      0.68      0.93      0.77      0.61     17205



In [21]:
# List the features sorted in descending order by feature importance
importances = brf.feature_importances_
indices = np.argsort(brf.feature_importances_)[::-1]
for f in range(X.shape[1]):
    print(f"{X.columns.values[indices[f]]}: ({importances[indices[f]]})")

total_rec_prncp: (0.08232559188092942)
total_pymnt_inv: (0.07514935891577482)
total_rec_int: (0.05295780612873343)
last_pymnt_amnt: (0.05172193147881896)
total_pymnt: (0.050666329293744046)
int_rate: (0.03180336578184677)
issue_d_Jan-2019: (0.02875447041125705)
mo_sin_old_rev_tl_op: (0.019698414607869152)
dti: (0.01785095742507862)
max_bal_bc: (0.016491603787959136)
revol_bal: (0.01582624735925119)
avg_cur_bal: (0.014814688214490035)
total_rev_hi_lim: (0.014568061942319082)
mths_since_recent_inq: (0.014059504036431206)
total_bc_limit: (0.01401471133549523)
total_bal_il: (0.013877798057673172)
out_prncp_inv: (0.013859222062267466)
bc_util: (0.013598791397331165)
mths_since_rcnt_il: (0.013429179784553854)
mo_sin_old_il_acct: (0.013423121714586898)
tot_hi_cred_lim: (0.013329292052855997)
installment: (0.013119605427149511)
annual_inc: (0.01310297280889863)
total_il_high_credit_limit: (0.012774095186915899)
bc_open_to_buy: (0.012708146158676319)
out_prncp: (0.012670087415701478)
tot_cur_ba

### Easy Ensemble Classifier

In [22]:
from imblearn.ensemble import EasyEnsembleClassifier

# Train the Classifier


In [23]:
# Calculated the balanced accuracy score
brf_easy = EasyEnsembleClassifier(100, random_state = 1)
brf_easy.fit(X_train_scaled,y_train)

EasyEnsembleClassifier(n_estimators=100, random_state=1)

In [24]:
y_predicted = brf_easy.predict(X_test_scaled)

In [25]:
# Display the confusion matrix
confusion_matrix(y_test, y_predicted)

array([[   85,     7],
       [ 1116, 15997]])

In [26]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_predicted))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.07      0.92      0.93      0.13      0.93      0.86        92
   low_risk       1.00      0.93      0.92      0.97      0.93      0.86     17113

avg / total       0.99      0.93      0.92      0.96      0.93      0.86     17205



In [27]:
# Calculated the balanced accuracy score
balanced_accuracy_score(y_test, y_predicted)*100

92.93497315796027

### Final Questions

1. Which model had the best balanced accuracy score?

    Ensemble has the better score compared to Rando Forest.

2. Which model had the best recall score?

    Ensemble has the better score compared to Rando Forest.

3. Which model had the best geometric mean score?

    YEnsemble has the better score compared to Rando Forest.

4. What are the top three features?

###     
    1) total_rec_prncp: (0.08232559188092942)
    2) total_pymnt_inv: (0.07514935891577482)
    3) total_rec_int: (0.05295780612873343)
    